FAQ: add MPI best practices (#1546)
hpc/charliecloud: doc/best_practices.rst

     and workload manager. That is, something like Process Management
     Interface (PMI) is needed to facilitate the job launch.

  3. **Shared memory**. Processes in separate sibling containers cannot use
     single-copy *cross-memory attach* (CMA), as opposed to double-copy POSIX
     or SysV shared memory. The solution is to put all ranks in the *same*
     container with :code:`ch-run --join`. (See above for details:
     :ref:`faq_join`.)

  4. **Network fabric.** Performant MPI jobs must recognize and use a system’s
     high-speed interconnect. Common issues that arise are:

       a. Libraries required to use the interconnect are proprietary or
          otherwise unavailable to the container.

       b. The interconnect is not supported by the container MPI.

     In both cases, the containerized MPI application will either fail or run
     significantly slower.

These problems can be avoided, and this section describes our recommendations
to do so.

Recommendations TL;DR
---------------------

Generally, we recommend building a flexible MPI container using:

   a. **libfabric** to flexibly manage process communication over a diverse
      set of network fabrics;

>    b. a parallel **process management interface** (PMI) compatible with the

Author: ```suggestion
   b. a parallel **process management interface** (PMI), e.g., pmix, pmi2,
      flux-pmi, etc., compatible with the host workload manager; and
```

The remaining sections detail the reasoning behind our approach. We recommend
referencing, or directly using, our examples
:code:`examples/Dockerfile.{libfabric,mpich,openmpi}`.

Use libfabric
-------------

`libfabric <https://ofiwg.github.io/libfabric>`_ (a.k.a. Open Fabrics
Interfaces or OFI) is a low-level communication library that abstracts diverse
networking technologies. It defines *providers* that implement the mapping
between application-facing software (e.g., MPI) and network specific drivers,
protocols, and hardware. These providers have been co-designed with fabric
hardware and application developers with a focus on HPC needs. libfabric lets
us more easily manage MPI communication over diverse network high-speed
interconnects (a.k.a. *fabrics*).

From our libfabric example (:code:`examples/Dockerfile.libfabric`):

.. literalinclude:: ../examples/Dockerfile.libfabric
   :language: docker
   :lines: 116-135

The above compiles libfabric with several “built-in” providers, i.e.
:code:`psm3` (on x86-64), :code:`rxm`, :code:`shm`, :code:`tcp`, and
:code:`verbs`, which enables MPI applications to run efficiently over most
verb devices using TCP, IB, OPA, and RoCE protocols.

> Two key advantages of using libfabric are: (1) the container’s libfabric can
> make use of “external” i.e. dynamic-shared-object (DSO) providers, and
> (2) Cray’s Slingshot provider (CXI) can be used by replacing the container
> image’s :code:`libfabric.so` with the Cray host’s.

Author: ```suggestion
Two key advantages of using libfabric are: (1) the container’s libfabric can
make use of “external”, i.e. dynamic-shared-object (DSO) providers, and
(2) libfabric replacement is much simpler than MPI replacement and preserves
the original container MPI, i.e., no host/container MPI ABI compatibility
issues; rather, we manage the more flexible libfabirc ABI compatibility. 
```

A DSO provider can be used by a libfabric that did not originally compile it,
i.e., they can be compiled on a target host and later injected into the
container along with any missing shared library dependencies, and used by the
container's libfabric. To build a libfabric provider as a DSO, add :code:`=dl`
to its :code:`configure` argument, e.g., :code:`--with-cxi=dl`.

A container's libfabric can also be replaced by a host libfabric. This is a
brittle but usually effective way to give containers access to the Cray
libfabric Slingshot provider :code:`cxi`.

In Charliecloud, both of these injection operations are currently done with
:code:`ch-fromhost`, though see `issue #1861
<https://github.com/hpc/charliecloud/issues/1861>`_.

Choose a compatible PMI
-----------------------

Unprivileged processes, including unprivileged containerized processes, are
unable to independently launch containerized processes on different nodes,

> aside from using SSH, which isn’t scalable. We must either (1)_rely on a host

Author: ```suggestion
aside from using SSH, which isn’t scalable. We must either (1) rely on a host
```

Reviewer: Thanks, this was supposed to be a non-breaking space, which shows up in my editor as a blue underscore.

> supported parallel process management interface (PMI), or (2)_achieve

Author: ```suggestion
supported parallel process management interface (PMI), or (2) achieve
```
